### PR TITLE
rtl8720dn: add support for rtl8720dn

### PR DESCRIPTION
**Pull request by sago35**
Created a driver for the RTL8720DN on the Seeed Wio Terminal.
In this PR, only TCP is supported.
UDP is also partially implemented, but no examples have been created yet.

The firmware of the RTL8720DN needs to be a newer version than 2.1.2.

* https://wiki.seeedstudio.com/Wio-Terminal-Network-Overview/

Seeed's firmware for RTL8720DN uses eRPC.
I first created a tool (go-erpcgen) to output the go code from eRPC IDL.
You can find it below, but it's not finished yet.
It's at least complete enough to make a TCP connection, but the stuff in `rtl8270dn/tmp.go` needs some more consideration.

* https://github.com/EmbeddedRPC/eRPC
* https://github.com/Seeed-Studio/seeed-ambd-firmware
* https://github.com/sago35/go-erpcgen

In `examples/tlsclient`, it accesses `www.example.com` instead of `tinygo.org`.
It could not communicate with `tinygo.org` with tls.
This is probably because the key length of `tinygo.org` is too long to communicate.

### CODE REVIEW

**Comment by sago35** [comment]
Can someone please review it for me?

**Comment by bgould** [comment]
I've been looking over the code for a few days, was impressed that you went so far as to write that tool (which I think is useful in its own right).  I'll try running the examples and also I'll take a closer look at the code overall

File: rtl8720dn/util.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	if 0 <= idx {
+		_, err := fmt.Sscanf(string(h[idx+16:]), "%d", &contentLength)
+		if err != nil {
+			return -1
+		}
+	}
+	return contentLength
+}
+
+// TODO: IPAddress implementation should be moved under drivers/net
```
Comment (deadprogram) [inline rtl8720dn/util.go:25]: :+1:

File: rtl8720dn/wifi.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
+package rtl8720dn
+
+import (
+	"fmt"
+	"time"
+)
+
+func (r *RTL8720DN) ConnectToAP(ssid string, password string) error {
```
Comment (deadprogram) [inline rtl8720dn/wifi.go:8]: In a future iteration we should create a interface for `wificlient` and another for `wifiaccesspoint` to have a uniform way to switch between this and wifinina and espat.

File: rtl8720dn/rpc_util.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+		if r.debug {
+			fmt.Printf("rx : %2d : ", n)
+			dumpHex(readBuf[:n])
+			fmt.Printf("\r\n")
+		}
+
+		length := uint16(readBuf[0]) + uint16(readBuf[1])<<8
+		crc := uint16(readBuf[2]) + uint16(readBuf[3])<<8
+		//fmt.Printf("len %d (%X) crc %04X\r\n", length, length, crc)
```
Comment (deadprogram) [inline rtl8720dn/rpc_util.go]: Should these commented out lines be removed or perhaps inside of `if r.debug` statements?

**Comment by deadprogram** [comment]
@sago35 also please resolve merge conflicts.

**Comment by bgould** [comment]
@sago35 I'm not able to get the webclient example to work on my wioterminal; when I run with debug turned on, this is all the output I see, it just seems to hang:
```
rpc_tcpip_adapter_init()
tx :  4 : 08 00 DE 26
tx :  8 : 00 01 0F 01 01 00 00 00
```

**Comment by sago35** [comment]
@bgould  Since rpc_tcpip_adapter_init() is the first UART process after power pin initialization, I expect it to work regardless of the environment.
I want you to check the firmware version of rtl8720dn.

> The firmware of the RTL8720DN needs to be a newer version than 2.1.2.
> https://wiki.seeedstudio.com/Wio-Terminal-Network-Overview/

You can also check if the following code works on the Arduino.
https://wiki.seeedstudio.com/Wio-Terminal-Wi-Fi/#faqs

```
#include "rpcWiFi.h"
 
void setup() {
    Serial.begin(115200);
    while(!Serial); // Wait to open Serial Monitor
    Serial.printf("RTL8720 Firmware Version: %s", rpc_system_version());
}
 
void loop() {
}
```



If the webclient works properly, you will get a response like the following, for example, and then proceed to rpc_wifi_off().

```
rpc_tcpip_adapter_init()
tx :  4 : 08 00 DE 26
tx :  8 : 00 01 0F 01 01 00 00 00
rx :  4 : 0C 00 A8 68
rx : 12 : 02 01 0F 01 01 00 00 00 00 00 00 00
rpc_wifi_off()
tx :  4 : 08 00 D3 25
tx :  8 : 00 1C 0E 01 02 00 00 00
rx :  4 : 0C 00 25 B1
rx : 12 : 02 1C 0E 01 02 00 00 00 00 00 00 00
```

File: rtl8720dn/wifi.go (hunk 2 above)
Comment (sago35) [reply rtl8720dn/wifi.go:8]: I have the same opinion.

**Comment by sago35** [comment]
To resolve conflicts in go.mod, it was necessary to rebase to resolve conflicts and then force-push.

**Comment by bgould** [comment]
I've been able to get the examples running, and in general the code looks good to me.

**Comment by deadprogram** [comment]
OK then let's merge and then future iterations can go from that point.

This is indeed some very impressive work @sago35 thank you for another amazing contribution. I think I will be using my WioTerminal a lot more now.

Merging.

**Pull request merged**